README.md
# Beaver Builder Class Dropdown
Adds user defined CSS classes to dropdown below the Beaver Builder class input in the Advanced tab

The settings for this plugin are located inside of Beaver Builder's settings, under the new Predefined Classes tab

![admin-panel](https://github.com/zackpyle/BBClassDropdown/assets/19413506/fa298823-8066-4215-ae22-f79b9aaac75b)


The groups and classes you define in the settings panel will then be displayed inside the builder in a dropdown below the class input in the Advanced tab

The little = icon next to the groups and classes let you drag/drop to reorder your groups and classes for your convenience. 

Below the predefined classes in the admin panel, there is a section to Import/Export Settings. It is useful when you have a default list of classes you like to use on every site (or as a starting point) - so you can create them once and import them into sites as you need.

The last and what I think is the **best** feature is you also have the ability to select if a group is a single class group - meaning, you only select one out of the group at a time. If you check this for the group, then in the dropdown, it will give you radio button looking symbols to let you know which class has been added aready. If you select another class, it will swap it for you. This is useful for things like background colors where you wouldn't have more than one of those classes at a time.

![frontend-dropdown](https://github.com/zackpyle/BBClassDropdown/assets/19413506/82d05c71-f675-4480-be2d-64e42924a1e4)


## Changelog

*v1.0.4 - 8/25/23*
- Remove select2 as it was causing a scrolling bug in the settings modal (issue #23)

*v1.0.3 - 8/24/23*
- Split up the plugin in functional classes and made it namespaced
- An autoloader takes care of loading/include_once/require_once the files
- Load admin script on fl-builder-settings page
- Fixed fatal error on plugins page (issue #19) - Updater fixed with wp_tempnam and unlink method
- Fixed reorder class bug (issue #21) - Couldn't reorder classes on new groups before hitting save

*v1.0.2 - 8/21/23*
- Fix error when opening other settings panels
- Moved all functions to BBClassDropdown

*v1.0.1 - 8/20/23*
- WP Updater connected so you can get the latest version pushed from Github straight to your WP install

*v0.7.0-beta - 8/19/23*
- Previously you could only clear existing classes/settings manually via a url query param. Now there is a new Reset button in the settings section at the bottom of the page
- After reset (and on initial activation), a new blank group is added so you don't have to click Add Group to get started

*v0.6.1-beta - 8/19/23*
- Fixed - v0.5.2 patch forgot to include rows and columns in the 'secondary modal' fix

*v0.6.0-beta - 8/19/23*
- New Import/Export Settings feature

*v0.5.2-beta - 8/18/23*
- Opening a secondary BB modal (like Post Module -> Custom Layout) would throw an error as the second modal assumed it was a module settings modal. Added more strict logic to look for the module settings modal

*v0.5.1-beta - 8/18/23*
- JS refactoring by badabingbreda

*v0.5.0-beta - 8/12/23*
- Drag and Drop reorder functionality for both Groups and Classes inside of Groups
- New table layout / styles
- Made the code more DRY
- First class or group can now be deleted if not only-child
- Don't need to confirm to delete an empty group
- Bug fixes

*v0.4.0-beta - 8/10/23*
- Feature - On groups with single selection substitute previous selected classname from optgroup with selection
- Feature - Fix for insertion of classes that are prefix of other classes. eg. if `uk-flex-center` already exists, you couln't add `uk-flex` later with callback handled by BB in (current) version 2.7.1.1
- Some code cleanup
- Renamed singleton checkbox from `checkbox` to `singleton` in all code.
- Select2 is enabled by default

*v0.3.3.0-beta - 8/7/23*
- Bug - SVG that holds symbols is taking up space on other tabs
- Bug - Typo when linking to Select2

*v0.3.2-beta - 8/7/23*
- Get rid of Remove Group on first group since you will never delete the first group

*v0.3.1-beta - 8/7/23*
- Bug fixes with disabled options
- Move Delete Group to same column as Delete Class

*v0.3.0-beta - 8/7/23*
- Gray out (disable) already used classes in the dropdown
- Add Select2 option

*v0.2.0-beta - 8/7/23*
- Settings page moved to Beaver Builder settings tab

*v0.1.1-beta - 8/6/23*
- Settings page saves to WP options

*v0.1.0-beta - 8/6/23*
- Initial commit


changelog.md
*1.7 - 8/19/23*
- Previously you could only clear existing classes/settings manually via a url query param. Now there is a new Reset button in the settings section at the bottom of the page
- After reset, new blank group is added so you don't have to click Add Group to get started

*1.6.1 - 8/19/23*
- Fixed - v1.5.2 patch forgot to include rows and columns in the secondary modal fix

*1.6 - 8/19/23*
- New Import/Export Settings feature

*1.5.2 - 8/18/23*
- Opening a secondary BB modal (like Post Module -> Custom Layout) would throw an error as the second modal assumed it was a module settings modal. Added more strict logic to look for the module settings modal

*1.5.1 - 8/18/23*
- JS refactoring by badabingbreda

*1.5 - 8/12/23*
- Drag and Drop reorder functionality for both Groups and Classes inside of Groups
- New table layout / styles
- Made the code more DRY
- First class or group can now be deleted if not only-child
- Don't need to confirm to delete an empty group
- Bug fixes

*1.4.0 - 8/10/23*
- Feature - On groups with single selection substitute previous selected classname from optgroup with selection
- Feature - Fix for insertion of classes that are prefix of other classes. eg. if `uk-flex-center` already exists, you couln't add `uk-flex` later with callback handled by BB in (current) version 2.7.1.1
- Some code cleanup
- Renamed singleton checkbox from `checkbox` to `singleton` in all code.
- Select2 is enabled by default

*1.3.3 - 8/7/23*
- Bug - SVG that holds symbols is taking up space on other tabs
- Bug - Typo when linking to Select2

*1.3.2 - 8/7/23*
- Get rid of Remove Group on first group since you will never delete the first group

*1.3.1 - 8/7/23*
- Bug fixes with disabled options
- Move Delete Group to same column as Delete Class

*1.3 - 8/7/23*
- Gray out (disable) already used classes in the dropdown
- Add Select2 option

*1.2 - 8/7/23*
- Settings page moved to Beaver Builder settings tab

*1.1 - 8/6/23*
- Settings page saves to WP options

*1.0 - 8/6/23*
- Initial commit


includes/bb-class-dropdown-admin.php
<?php

/**
 * Provide a admin area for the plugin
 */


// Exit if accessed directly
if ( ! defined( 'ABSPATH' ) ) {
    exit;
}

/**
 * Settings page HTML
 */
function beaver_builder_class_dropdown_settings_page_html() {
    // Get options
    $options = get_option( 'beaver_builder_class_dropdown_options', array() );
    // Render settings page
    ?>
	
    <div id="fl-class-dropdown-form" class="fl-settings-form">
        <h1>Predefined Classes</h1>
        <form action="" method="post" id="class-dropdown-form">
			<input type="hidden" name="bb-class-dd-nonce" value="<?php echo wp_create_nonce('bb-class-dd-nonce'); ?>">
			<input type="hidden" name="bb-class-action" value="update">
            <?php settings_fields( 'beaver_builder_class_dropdown_options_group' ); ?>
            <?php do_settings_sections( 'beaver_builder_class_dropdown_options_group' ); ?>
			<table class="beaver-builder-class-dropdown-groups">
				<thead>
					<tr>
						<th id="handle-th"></th> <!-- Empty th for the handle -->
						<th id="group-name-th" scope="row"><?php esc_html_e( 'Group', 'BBClassDropdown' ); ?></th>
						<th id="class-th"><?php esc_html_e( 'Class / Label', 'BBClassDropdown' ); ?></th>
						<th id="button-th"></th> <!-- Empty th for the buttons -->
					</tr>
				</thead>
				<tbody>
					<?php 
	if (isset($options['groups']) && is_array($options['groups'])) {
		foreach ( $options['groups'] as $i => $group ) :
		if(is_array($group)) { // Add this line to check if $group is an array
					?>
					<tr class="group">
						<td class="group-handle"  valign="top">
							<!-- Drag-and-drop handle -->
							<div class="drag-handle"></div> 
							<!-- Hidden input for ordering -->
              						<input class="group-order" type="hidden" data-name="order" value="<?php echo $i; ?>" />
						</td>
						<td valign="top" class="group-name-col">
							<input class="group-name" type="text" data-name="name" value="<?php echo array_key_exists('name', $group) ? esc_attr( $group['name'] ) : ""; ?>" />
							<div class="group-options">
								<button type="button" class="button beaver-builder-class-dropdown-remove-group">
									<span class="sr-only">Delete Group</span><svg aria-hidden="true" width="13" height="13"><use xlink:href="#trash" /></svg>
								</button>
								<!-- Option to have group be only a single class at a time - useful for something like a background color -->
								<div class="single-select-wrapper">
									<label><input type="checkbox" data-name="singleton" value="1" <?php checked( isset($group['singleton']) ? $group['singleton'] : 0 ); ?> /> Single Class Group</label>
								</div>
							</div>
						</td>
						<td valign="top" class="class-col">
							<table class="beaver-builder-class-dropdown-classes">
								<tbody>
									<?php 
						if (isset($group['classes']) && is_array($group['classes'])) {
							foreach ( $group['classes'] as $j => $class ) :
							if(is_array($class)) { // Add this line to check if $class is an array
									?>
									<tr class="class-row" valign="top">
										<td class="class-handle" valign="top">
											<!-- Drag-and-drop handle -->
											<div class="drag-handle"></div>
											<!-- Hidden input for ordering -->
											<input class="class-order" type="hidden" data-name="order" value="<?php echo $j; ?>" />
										</td>
										<td class="class-value-col"><input type="text" data-name="id" placeholder="foo-bar" value="<?php echo array_key_exists('id', $class) ? esc_attr( $class['id'] ) : ""; ?>" /></td>
										<td class="class-label-col"><input type="text" data-name="name" placeholder="Foo Bar" value="<?php echo array_key_exists('name', $class) ? esc_attr( $class['name'] ) : ""; ?>" /></td>
										<td class="class-btn-col">
											<button type="button" class="button beaver-builder-class-dropdown-remove-class">
											<svg aria-hidden="true" width="13" height="13">
												<use xlink:href="#trash" />
											</svg>
											</button>
										</td>
									</tr>
									
									<?php
							} // Closing bracket for if(is_array($class))
							endforeach;
						}
									?>
								</tbody>
							</table>
							<button type="button" class="button beaver-builder-class-dropdown-add-class">
								<svg aria-hidden="true" width="12" height="12">
									<use xlink:href="#plus" />
								</svg>
							</button>
						</td>
					</tr>
					<?php
		} // Closing bracket for if(is_array($group))
		endforeach;
	}
					?>
				</tbody>
			</table>
			<button type="button" class="button beaver-builder-class-dropdown-add-group">Add Group</button>
            <div id="main-submit-wrapper">
				<?php submit_button(); ?>
			</div>
        </form>
		<section id="bb-class-dd-settings">
			<div id="bb-class-import-export-settings">
				<h2>Import/Export Settings</h2>
				<button id="export-classes" class="button button-primary">Export Classes</button>
				<button id="import-classes" class="button button-primary">Import Classes</button>
				<dialog id="import-modal" onclick="event.target==this && this.close()">
					<button id="close-import-modal" aria-label="Close Import Dialog">
						<span aria-hidden="true">&times;</span>
					</button>
					<div class="modal-content">
						<h3>Select JSON import file:</h3>
						<form action="" method="post" enctype="multipart/form-data" id="class-import-form">
							<input type="hidden" name="bb-class-dd-nonce" value="<?php echo wp_create_nonce('bb-class-dd-nonce'); ?>">
							<input type="hidden" name="bb-class-action" value="import">
							<?php settings_fields( 'beaver_builder_class_dropdown_options_group' ); ?>
							<?php do_settings_sections( 'beaver_builder_class_dropdown_options_group' ); ?>
							<input type="file" name="fileToUpload" id="fileToUpload" accept="application/JSON">
							<?php submit_button( 'Import Settings' ); ?>
						</form>
					</div>
				</dialog>
			</div>
			<div id="bb-dropdown-reset-wrapper">
				<h2>Reset Settings</h2>
				<button id="reset-bb-dropdown-settings" class="button-link-delete button-secondary">Delete All Classes</button>
			</div>
			
		</section>
	</div>

	<svg xmlns="http://www.w3.org/2000/svg"  hidden id="icon-svg-container">
		<symbol id="trash" viewBox="0 0 448 512" >
			<path d="M170.5 51.6L151.5 80h145l-19-28.4c-1.5-2.2-4-3.6-6.7-3.6H177.1c-2.7 0-5.2 1.3-6.7 3.6zm147-26.6L354.2 80H368h48 8c13.3 0 24 10.7 24 24s-10.7 24-24 24h-8V432c0 44.2-35.8 80-80 80H112c-44.2 0-80-35.8-80-80V128H24c-13.3 0-24-10.7-24-24S10.7 80 24 80h8H80 93.8l36.7-55.1C140.9 9.4 158.4 0 177.1 0h93.7c18.7 0 36.2 9.4 46.6 24.9zM80 128V432c0 17.7 14.3 32 32 32H336c17.7 0 32-14.3 32-32V128H80zm80 64V400c0 8.8-7.2 16-16 16s-16-7.2-16-16V192c0-8.8 7.2-16 16-16s16 7.2 16 16zm80 0V400c0 8.8-7.2 16-16 16s-16-7.2-16-16V192c0-8.8 7.2-16 16-16s16 7.2 16 16zm80 0V400c0 8.8-7.2 16-16 16s-16-7.2-16-16V192c0-8.8 7.2-16 16-16s16 7.2 16 16z"/>
		</symbol>
		<symbol id="plus" viewBox="0 0 448 512" >
			<path d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"/>
		</symbol>
	</svg>
	<style>
		.fl-settings-form{
			max-width: 700px;
		}
		table{
			width:100%;
		}
		input.group-name,
		.class-row input{
			width: 100%;
		}
		td.class-btn-col {
			min-width: 25px;
			padding-left: 5px !important;
		}
		#fl-class-dropdown-form input::placeholder {
			color: #b7b7b7;
		}
		.beaver-builder-class-dropdown-groups{
			margin-top: 20px;
			border-collapse: collapse;
		}
		.beaver-builder-class-dropdown-groups tr.group:not(:first-child) > td {
			padding-top: 20px;
		}
		.beaver-builder-class-dropdown-groups th{
			text-align: left;
			padding-bottom:5px !important;
		}
		.beaver-builder-class-dropdown-classes{
			margin-top: -2px;
			margin-left:10px;
		}
		tr.group {
			border-bottom: 1px solid #f2f2f2;
		}
		tr.group:last-child{
			border-bottom:none;
		}
		button.beaver-builder-class-dropdown-add-group{
			margin-left: 32px !important;
		}
		button.button.beaver-builder-class-dropdown-remove-class,
		button.button.beaver-builder-class-dropdown-remove-group{
			border-radius: 50%;
			height: 25px;
			width: 25px;
			display: grid;
			place-content: center;
			min-height: unset;
			margin-top: 5px;
			padding: 0;
		}
		button.button.beaver-builder-class-dropdown-remove-class svg,
		button.button.beaver-builder-class-dropdown-add-class svg,
		button.beaver-builder-class-dropdown-remove-group svg{
			fill: currentColor;
		}
		.beaver-builder-class-dropdown-groups button.beaver-builder-class-dropdown-add-class {
			margin: 5px 0 20px 50px;
			border-radius: 50%;
			height: 20px;
			width: 20px;
			display: grid;
			place-content: center;
			min-height: unset;
			padding: 0;
		}
		.beaver-builder-class-dropdown-groups .group:only-child .beaver-builder-class-dropdown-remove-group,
		.beaver-builder-class-dropdown-classes tr:only-child .beaver-builder-class-dropdown-remove-class {
			display: none;
		}
		.fl-settings-form .beaver-builder-class-dropdown-groups th{
			padding: 0;
			font-weight: bold
		}
		.fl-settings-form .beaver-builder-class-dropdown-groups td {
			padding: 0;
		}
		#select2-settings{
			margin-top: 60px;
		}
		.single-select-wrapper{
			margin-top:10px;
		}
		#icon-svg-container{
			border: 0 !important;
			clip: rect(1px, 1px, 1px, 1px) !important; /* 1 */
			-webkit-clip-path: inset(50%) !important;
				clip-path: inset(50%) !important;  /* 2 */
			height: 1px !important;
			margin: -1px !important;
			overflow: hidden !important;
			padding: 0 !important;
			position: absolute !important;
			width: 1px !important;
			white-space: nowrap !important;
		}
		/* Drag Handle */
		.drag-handle {
			position: relative;
			width: 15px;
			cursor: grab;
			padding: 10px 3px 10px 10px;
		}
		.drag-handle::before,
		.drag-handle::after {
			content: '';
			display: block;
			width: 100%;
			height: 2px;
			background-color: #cecece;
			margin-bottom: 3px;
		}
		.drag-handle::after {
			margin-bottom: 0px;
		}
		th#class-th {
			padding-left: 45px;
		}
		.group-options {
			display: flex;
			align-items: center;
			gap: 15px;
			margin-top: 5px;
		}
		.beaver-builder-class-dropdown-groups .group:first-child .group-options{
			margin-top:0;
		}
		.sr-only {
			border: 0 !important;
			clip: rect(1px, 1px, 1px, 1px) !important;
			-webkit-clip-path: inset(50%) !important;
				clip-path: inset(50%) !important;
			height: 1px !important;
			margin: -1px !important;
			overflow: hidden !important;
			padding: 0 !important;
			position: absolute !important;
			width: 1px !important;
			white-space: nowrap !important;
		}
		#main-submit-wrapper .submit{
			text-align:right;
		}
		.wp-core-ui .button-link-delete.button-secondary {
			border-color: currentColor;
		}
		#bb-dropdown-reset-wrapper{
			margin-top:30px;
		}
		#bb-class-dd-settings h2{
			font-weight:500;
		}
		
		/* Import Export */		
		.modal-content{
			padding:35px;
		}
		#bb-class-dd-settings{
			margin-top:80px;
		}
		#import-modal{
			border: none;
			border-radius:5px;
			box-shadow: 0 5px 10px rgb(0 0 0 / .25);
			overflow:visible;
			padding:0;
		}
		#import-modal::backdrop {
		  background: rgb(0 0 0 / 0.4);
		}
		button#close-import-modal {
			position: absolute;
			top: -15px;
			right: -15px;
			height: 30px;
			width: 30px;
			border-radius: 50%;
			border: 1px solid gray;
			display: grid;
			place-content: center;
			line-height: 1;
			cursor: pointer;
			font-size: 17px;
		}
		#import-modal p.submit{
			margin-top:0;
			padding-bottom:0;
		}
	</style>
    <?php
}

function bb_class_dropdown_admin_settings_save() {
	// return early when not set or nonce not matching

    if ( !isset($_POST['bb-class-dd-nonce']) || !wp_verify_nonce($_POST['bb-class-dd-nonce'], 'bb-class-dd-nonce') ) return;
	
	$action = filter_input( INPUT_POST , 'bb-class-action' );

	if ( 'update' == $action ) {

		$settings = isset($_POST['beaver_builder_class_dropdown_options']) ? $_POST['beaver_builder_class_dropdown_options'] : array();

		$group_order = 0;
		// Sanitize each group and class setting
		if (isset($settings['groups']) && is_array($settings['groups'])) {
	
			foreach ($settings['groups'] as $group) {
	
				// add our new group
				$__new_groups[ $group_order ] = [];
				
				// set a name for this new group
				$__new_groups[ $group_order ]['name'] = isset( $group['name'] ) ? sanitize_text_field( $group['name'] ) : "";
	
				if (isset($group['classes']) && is_array($group['classes'])) {
	
					// clear out our list of classes for this group
					$classes = [];
					
					// start class order at 0
					$class_order = 0;
					foreach ($group['classes'] as $class) {
	
						$classes[] = array(
									'id' => isset($class['id']) ? sanitize_text_field($class['id']) : "",
									'name' => isset($class['name']) ? sanitize_text_field($class['name']) : "",
									'order' => $class_order,
						);
						// increment class_order
						$class_order++;
					}
	
					// Replace the associative classes array with the sequential one
					$__new_groups[ $group_order ]['classes'] = $classes;
					$__new_groups[ $group_order ][ 'singleton' ] = isset( $group[ 'singleton' ] ) ? '1' : '0';
				}
				$__new_groups[ $group_order ][ 'order' ] = $group_order;
				
				// increment group_order
				$group_order++;
			}
	
			$settings[ 'groups' ] = $__new_groups;

			// Save to db
			update_option('beaver_builder_class_dropdown_options', $settings);
			
		}
	
	} elseif ( 'import' == $action ) {

		if ( isset($_FILES["fileToUpload"] ) && $_FILES[ 'fileToUpload' ][ 'tmp_name' ] !== '' ) {
			$content = file_get_contents( $_FILES["fileToUpload"]["tmp_name"] );

			try {
				$json = json_decode( $content , true );
			} catch ( \Exception $error) {
				return false;
			}
			// Save to db
			update_option('beaver_builder_class_dropdown_options', $json);

		}

	}



	FLBuilderModel::delete_asset_cache_for_all_posts();
    
}


add_action( 'init', 'bb_class_dropdown_admin_settings_save' );
